File: apps/backend/src/services/streaks.ts
import { randomBytes } from "node:crypto"
import { and, asc, desc, eq, isNull, sql } from "drizzle-orm"
import { db } from "@yomi/db"
import * as authSchema from "../auth-schema.js"

const LEADERBOARD_LIMIT = 50

const HANDLE_ADJECTIVES = [
  "quiet",
  "swift",
  "brave",
  "calm",
  "bright",
  "bold",
  "gentle",
  "clever",
  "steady",
  "quick",
  "sharp",
  "keen",
  "wise",
  "eager",
  "vivid",
  "nimble",
]
const HANDLE_NOUNS = [
  "falcon",
  "otter",
  "maple",
  "comet",
  "ember",
  "harbor",
  "willow",
  "granite",
  "cedar",
  "raven",
  "meadow",
  "quartz",
  "lynx",
  "aspen",
  "delta",
  "orbit",
]

const HANDLE_PATTERN = /^[a-z][a-z0-9-]{2,23}$/

function isDuplicateHandleError(err: unknown): boolean {
  return String(err).includes("user_leaderboard_handle_unique")
}

function generateHandle(): string {
  const adjective = HANDLE_ADJECTIVES[Math.floor(Math.random() * HANDLE_ADJECTIVES.length)]
  const noun = HANDLE_NOUNS[Math.floor(Math.random() * HANDLE_NOUNS.length)]
  const suffix = randomBytes(2).toString("hex")
  return `${adjective}-${noun}-${suffix}`
}

function avatarUrlFor(row: { id: string; image: string | null; customAvatarKey: string | null }) {
  if (row.customAvatarKey) return `/api/user/avatar/${row.id}`
  return row.image ?? null
}

function todayUtc(): string {
  return new Date().toISOString().slice(0, 10)
}

function yesterdayUtc(from: string): string {
  const d = new Date(`${from}T00:00:00.000Z`)
  d.setUTCDate(d.getUTCDate() - 1)
  return d.toISOString().slice(0, 10)
}

// Generates and persists a unique leaderboard handle for a user who doesn't
// have one yet. Called both from recordDailyActivity (a user's first-ever
// message) and the one-off backfill script for pre-existing users.
export async function ensureLeaderboardHandle(userId: string): Promise<string> {
  for (let attempt = 0; attempt < 3; attempt++) {
    const candidate = generateHandle()
    try {
      await db
        .update(authSchema.user)
        .set({ leaderboardHandle: candidate })
        .where(eq(authSchema.user.id, userId))
      return candidate
    } catch (err) {
      if (!isDuplicateHandleError(err)) throw err
    }
  }
  throw new Error("failed to generate a unique leaderboard handle after 3 attempts")
}

export async function getAvatarKey(userId: string): Promise<string | null> {
  const [row] = await db
    .select({ customAvatarKey: authSchema.user.customAvatarKey })
    .from(authSchema.user)
    .where(eq(authSchema.user.id, userId))
    .limit(1)
  return row?.customAvatarKey ?? null
}

export async function recordDailyActivity(userId: string): Promise<void> {
  const [row] = await db
    .select({
      currentStreak: authSchema.user.currentStreak,
      longestStreak: authSchema.user.longestStreak,
      lastActiveDate: authSchema.user.lastActiveDate,
      leaderboardHandle: authSchema.user.leaderboardHandle,
    })
    .from(authSchema.user)
    .where(eq(authSchema.user.id, userId))
    .limit(1)
  if (!row) return

  if (!row.leaderboardHandle) {
    await ensureLeaderboardHandle(userId)
  }

  const today = todayUtc()
  if (row.lastActiveDate === today) {
    await db
      .update(authSchema.user)
      .set({ totalMessagesSent: sql`${authSchema.user.totalMessagesSent} + 1` })
      .where(eq(authSchema.user.id, userId))
    return
  }

  const wasYesterday = row.lastActiveDate === yesterdayUtc(today)
  const newStreak = wasYesterday ? row.currentStreak + 1 : 1
  const newLongest = Math.max(row.longestStreak, newStreak)

  await db
    .update(authSchema.user)
    .set({
      totalMessagesSent: sql`${authSchema.user.totalMessagesSent} + 1`,
      currentStreak: newStreak,
      longestStreak: newLongest,
      lastActiveDate: today,
    })
    .where(eq(authSchema.user.id, userId))
}

export async function getStreakStats(userId: string): Promise<{
  currentStreak: number
  longestStreak: number
  totalMessagesSent: number
  leaderboardOptIn: boolean
  leaderboardHandle: string | null
  leaderboardShowPhoto: boolean
  avatarUrl: string | null
  plan: string
}> {
  const [row] = await db
    .select({
      currentStreak: authSchema.user.currentStreak,
      longestStreak: authSchema.user.longestStreak,
      totalMessagesSent: authSchema.user.totalMessagesSent,
      leaderboardOptIn: authSchema.user.leaderboardOptIn,
      leaderboardHandle: authSchema.user.leaderboardHandle,
      leaderboardShowPhoto: authSchema.user.leaderboardShowPhoto,
      image: authSchema.user.image,
      customAvatarKey: authSchema.user.customAvatarKey,
      plan: authSchema.user.plan,
    })
    .from(authSchema.user)
    .where(eq(authSchema.user.id, userId))
    .limit(1)
  if (!row) {
    return {
      currentStreak: 0,
      longestStreak: 0,
      totalMessagesSent: 0,
      leaderboardOptIn: false,
      leaderboardHandle: null,
      leaderboardShowPhoto: true,
      avatarUrl: null,
      plan: "explore",
    }
  }
  return {
    currentStreak: row.currentStreak,
    longestStreak: row.longestStreak,
    totalMessagesSent: row.totalMessagesSent,
    leaderboardOptIn: row.leaderboardOptIn,
    leaderboardHandle: row.leaderboardHandle,
    leaderboardShowPhoto: row.leaderboardShowPhoto,
    avatarUrl: avatarUrlFor({ id: userId, image: row.image, customAvatarKey: row.customAvatarKey }),
    plan: row.plan,
  }
}

export async function setLeaderboardOptIn(
  userId: string,
  optIn: boolean,
): Promise<{ leaderboardOptIn: boolean; leaderboardHandle: string | null }> {
  const [row] = await db
    .select({ leaderboardHandle: authSchema.user.leaderboardHandle })
    .from(authSchema.user)
    .where(eq(authSchema.user.id, userId))
    .limit(1)

  await db
    .update(authSchema.user)
    .set({ leaderboardOptIn: optIn })
    .where(eq(authSchema.user.id, userId))

  return { leaderboardOptIn: optIn, leaderboardHandle: row?.leaderboardHandle ?? null }
}

export async function updateLeaderboardHandle(
  userId: string,
  handle: string,
): Promise<{ ok: true; leaderboardHandle: string } | { ok: false; error: string }> {
  const trimmed = handle.trim().toLowerCase()
  if (!HANDLE_PATTERN.test(trimmed)) {
    return {
      ok: false,
      error:
        "Handle must be 3-24 characters: lowercase letters, numbers, and dashes, starting with a letter.",
    }
  }
  try {
    await db
      .update(authSchema.user)
      .set({ leaderboardHandle: trimmed })
      .where(eq(authSchema.user.id, userId))
    return { ok: true, leaderboardHandle: trimmed }
  } catch (err) {
    if (isDuplicateHandleError(err)) {
      return { ok: false, error: "That handle is already taken — try another." }
    }
    throw err
  }
}

export async function setLeaderboardShowPhoto(
  userId: string,
  showPhoto: boolean,
): Promise<{ leaderboardShowPhoto: boolean }> {
  await db
    .update(authSchema.user)
    .set({ leaderboardShowPhoto: showPhoto })
    .where(eq(authSchema.user.id, userId))
  return { leaderboardShowPhoto: showPhoto }
}

export async function getLeaderboard(userId: string): Promise<{
  entries: {
    rank: number
    handle: string
    totalMessagesSent: number
    isYou: boolean
    avatarUrl: string | null
    plan: string
  }[]
  yourRank: number | null
}> {
  const top = await db
    .select({
      id: authSchema.user.id,
      handle: authSchema.user.leaderboardHandle,
      totalMessagesSent: authSchema.user.totalMessagesSent,
      image: authSchema.user.image,
      customAvatarKey: authSchema.user.customAvatarKey,
      leaderboardShowPhoto: authSchema.user.leaderboardShowPhoto,
      plan: authSchema.user.plan,
    })
    .from(authSchema.user)
    .where(and(eq(authSchema.user.leaderboardOptIn, true), isNull(authSchema.user.deletedAt)))
    .orderBy(desc(authSchema.user.totalMessagesSent), asc(authSchema.user.id))
    .limit(LEADERBOARD_LIMIT)

  const entries = top.map((row, index) => ({
    rank: index + 1,
    handle: row.handle ?? "anonymous",
    totalMessagesSent: row.totalMessagesSent,
    isYou: row.id === userId,
    avatarUrl: row.leaderboardShowPhoto
      ? avatarUrlFor({ id: row.id, image: row.image, customAvatarKey: row.customAvatarKey })
      : null,
    plan: row.plan,
  }))

  const inTop = entries.find((e) => e.isYou)
  if (inTop) return { entries, yourRank: inTop.rank }

  const [viewer] = await db
    .select({
      leaderboardOptIn: authSchema.user.leaderboardOptIn,
      totalMessagesSent: authSchema.user.totalMessagesSent,
    })
    .from(authSchema.user)
    .where(eq(authSchema.user.id, userId))
    .limit(1)

  if (!viewer?.leaderboardOptIn) return { entries, yourRank: null }

  const [countRow] = await db
    .select({ count: sql<number>`count(*)::int` })
    .from(authSchema.user)
    .where(
      and(
        eq(authSchema.user.leaderboardOptIn, true),
        isNull(authSchema.user.deletedAt),
        sql`${authSchema.user.totalMessagesSent} > ${viewer.totalMessagesSent}`,
      ),
    )
    .limit(1)

  return { entries, yourRank: (countRow?.count ?? 0) + 1 }
}

File: apps/backend/src/yomi/services/streaks.py
"""Leaderboard + daily-streak stats.

Port of apps/backend/src/services/streaks.ts.
"""

from __future__ import annotations

import re

from sqlalchemy import asc, desc, func, select, update
from sqlalchemy.ext.asyncio import AsyncSession

from yomi.db.models_auth import User

HANDLE_PATTERN = re.compile(r"^[a-z][a-z0-9-]{2,23}$")

LEADERBOARD_LIMIT = 50

HANDLE_ADJECTIVES = [
    "quiet",
    "swift",
    "brave",
    "calm",
    "bright",
    "bold",
    "gentle",
    "clever",
    "steady",
    "quick",
    "sharp",
    "keen",
    "wise",
    "eager",
    "vivid",
    "nimble",
]
HANDLE_NOUNS = [
    "falcon",
    "otter",
    "maple",
    "comet",
    "ember",
    "harbor",
    "willow",
    "granite",
    "cedar",
    "raven",
    "meadow",
    "quartz",
    "lynx",
    "aspen",
    "delta",
    "orbit",
]


def is_duplicate_handle_error(err: Exception) -> bool:
    return "user_leaderboard_handle_unique" in str(err)


async def update_leaderboard_handle(
    session: AsyncSession, user_id: str, handle: str
) -> dict[str, object]:
    """Returns {ok: True, leaderboardHandle} or {ok: False, error}."""
    trimmed = handle.strip().lower()
    if not HANDLE_PATTERN.fullmatch(trimmed):
        return {
            "ok": False,
            "error": (
                "Handle must be 3-24 characters: lowercase letters, numbers, and "
                "dashes, starting with a letter."
            ),
        }
    try:
        await session.execute(
            update(User).where(User.id == user_id).values(leaderboard_handle=trimmed)
        )
    except Exception as err:  # noqa: BLE001 — normalized below
        if is_duplicate_handle_error(err):  # type: ignore[arg-type]
            return {"ok": False, "error": "That handle is already taken — try another."}
        raise
    return {"ok": True, "leaderboardHandle": trimmed}


async def set_leaderboard_show_photo(
    session: AsyncSession, user_id: str, show_photo: bool
) -> dict[str, bool]:
    await session.execute(
        update(User).where(User.id == user_id).values(leaderboard_show_photo=show_photo)
    )
    return {"leaderboardShowPhoto": show_photo}


async def get_streak_stats(session: AsyncSession, user_id: str) -> dict[str, object]:
    row = (
        await session.execute(
            select(
                User.current_streak,
                User.longest_streak,
                User.total_messages_sent,
                User.leaderboard_opt_in,
                User.leaderboard_handle,
                User.leaderboard_show_photo,
                User.image,
                User.plan,
            ).where(User.id == user_id)
        )
    ).one_or_none()
    if row is None:
        return {
            "currentStreak": 0,
            "longestStreak": 0,
            "totalMessagesSent": 0,
            "leaderboardOptIn": False,
            "leaderboardHandle": None,
            "leaderboardShowPhoto": True,
            "avatarUrl": None,
            "plan": "explore",
        }
    return {
        "currentStreak": row.current_streak,
        "longestStreak": row.longest_streak,
        "totalMessagesSent": row.total_messages_sent,
        "leaderboardOptIn": row.leaderboard_opt_in,
        "leaderboardHandle": row.leaderboard_handle,
        "leaderboardShowPhoto": row.leaderboard_show_photo,
        "avatarUrl": row.image,
        "plan": row.plan,
    }


async def set_leaderboard_opt_in(
    session: AsyncSession, user_id: str, opt_in: bool
) -> dict[str, object]:
    handle = (
        await session.execute(
            select(User.leaderboard_handle).where(User.id == user_id).limit(1)
        )
    ).scalar_one_or_none()
    await session.execute(
        update(User).where(User.id == user_id).values(leaderboard_opt_in=opt_in)
    )
    return {"leaderboardOptIn": opt_in, "leaderboardHandle": handle}


async def get_leaderboard(session: AsyncSession, user_id: str) -> dict[str, object]:
    rows = (
        await session.execute(
            select(
                User.id,
                User.leaderboard_handle,
                User.total_messages_sent,
                User.image,
                User.leaderboard_show_photo,
                User.plan,
            )
            .where(User.leaderboard_opt_in.is_(True), User.deleted_at.is_(None))
            .order_by(desc(User.total_messages_sent), asc(User.id))
            .limit(LEADERBOARD_LIMIT)
        )
    ).all()

    entries = []
    for index, row in enumerate(rows):
        entries.append(
            {
                "rank": index + 1,
                "handle": row.leaderboard_handle or "anonymous",
                "totalMessagesSent": row.total_messages_sent,
                "isYou": row.id == user_id,
                "avatarUrl": row.image if row.leaderboard_show_photo else None,
                "plan": row.plan,
            }
        )

    in_top = next((e for e in entries if e["isYou"]), None)
    if in_top is not None:
        return {"entries": entries, "yourRank": in_top["rank"]}

    viewer = (
        await session.execute(
            select(User.leaderboard_opt_in, User.total_messages_sent)
            .where(User.id == user_id)
            .limit(1)
        )
    ).one_or_none()
    if viewer is None or not viewer.leaderboard_opt_in:
        return {"entries": entries, "yourRank": None}

    count = (
        await session.execute(
            select(func.count())
            .select_from(User)
            .where(
                User.leaderboard_opt_in.is_(True),
                User.deleted_at.is_(None),
                User.total_messages_sent > viewer.total_messages_sent,
            )
        )
    ).scalar_one()

    return {"entries": entries, "yourRank": count + 1}